### PR TITLE
sanitize python env and subprocess i/o

### PR DESCRIPTION
**Pull request by pmp-p**
While working with python 3.11 wasm and pip modules, emcc is often called as subprocess from a tweaked python.
 
i encountered some troubles, emcc.py is inheriting some unwanted values from parent shell
to name a few _PYTHON_SYSCONFIGDATA_NAME and PYTHONHOME / PYTHONPATH

adding -E will prevent PYTHON* to interfere, and the unset _PYTHON_SYSCONFIGDATA_NAME is very unlikely to be used anywhere except in cpython modules cross compiling.

### CODE REVIEW

**Comment by sbc100** [comment]
I'm a little worried about removing these from the environment since you can imagine some users wanting to set PYTHONPATH to point to a location where they have install some python modules (for example the ones in `requirements-dev.txt`).. but I suppose it might be good default and advanced users can bypass the launcher scripts maybe?

**Comment by pmp-p** [comment]
> but I suppose it might be good default and advanced users can bypass the launcher scripts maybe?

indeed it absolutely does not cover all cases, it's just sane default to provide basic module compiling for CPython 3.11+

for building pygame on wasm with https://github.com/pygame-web/python-wasm-sdk , emcc has finally turned to something more like :
```py
#!/bin/bash

unset _PYTHON_SYSCONFIGDATA_NAME
unset PYTHONHOME
unset PYTHONPATH

for arg do
    shift
    [ "$arg" = "-I/usr/include" ] && continue
    [ "$arg" = "-I/usr/include/SDL2" ] && continue
    [ "$arg" = "-L/usr/lib64" ]	&& continue
    [ "$arg" = "-L/usr/lib" ]   && continue
    set -- "$@" "$arg"
done

SHARED=""
IS_SHARED=false

if echo "$@"|grep -q shared
then
    IS_SHARED=true
    SHARED="-sSIDE_MODULE"
else
    if echo "$@"|grep -q wasm32-emscripten.so
    then
        IS_SHARED=true
        SHARED="-shared -sSIDE_MODULE"
    fi
fi

if $IS_SHARED
then
    $SYS_PYTHON -E $0.py  -L/opt/python-wasm-sdk/devices/emsdk/usr/lib -Wno-unused-command-line-argument "$@"
else
    $SYS_PYTHON -E $0.py -I/opt/python-wasm-sdk/devices/emsdk/usr/include -Wno-unused-command-line-argument "$@"
fi
```
where SYS_PYTHON is set to stable/system python ( can be version -1 when building alpha/beta )
via 
```sh
for py in 10 9 8 7
do
    if command -v python3.${py} >/dev/null
    then
        export SYS_PYTHON=$(command -v python3.${py})
        break
    fi
done
```
SYS_PYTHON was added because EMSDK_PYTHON is cleared by emsdk_env.sh ( why ? )

**Comment by hoodmane** [comment]
I think this change would break Pyodide's package compilation.

**Comment by hoodmane** [comment]
In particular, we need to explicitly control `_PYTHON_SYSCONFIGDATA_NAME` and other similar environment variables to set up the cross compilation environment. I think you should explicitly set up the environment you need and then `emcc` should pass it along. I think this is the way that environment variables normally work.

**Comment by sbc100** [comment]
> In particular, we need to explicitly control `_PYTHON_SYSCONFIGDATA_NAME` and other similar environment variables to set up the cross compilation environment. I think you should explicitly set up the environment you need and then `emcc` should pass it along. I think this is the way that environment variables normally work.

I don't think this change stops you from using `_PYTHON_SYSCONFIGDATA_NAME` does it?  It just ignore it while emcc is running.. it doesn't effect processes outside of the compiler itself.

You don't need that environment variable to be set within the compiler/emcc do you?

**Comment by hoodmane** [comment]
> You don't need that environment variable to be set within the compiler/emcc do you?

No, I guess not, we don't need them inside of the compiler wrapper. I guess this change actually makes a lot of sense because it defends `emcc.py` against the custom cross compilation Python environment set up outside of it.

File: tools/shared.py [hunk 1]
```diff
@@ -103,6 +103,7 @@ def run_process(cmd, check=True, input=None, *args, **kw):
   sys.stdout.flush()
   sys.stderr.flush()
   kw.setdefault('universal_newlines', True)
+  kw.setdefault('encoding', 'utf-8')
```
[the remark this answers is not in the document]
Comment (pmp-p) [reply tools/shared.py:106]: iirc on python 3.6 / 3.7 only. Anyway ascii is no more a valid encoding for subprocesses so it's safe to enforce utf-8.

[the remark this answers is not in the document]
Comment (sbc100) [reply tools/shared.py:106]: I'd rather separate this part of the change, and I would rather have an example or test that fails without it before landing.   Can you split this out from the wrapper script changes?

File: em++ [hunk 2]
```diff
@@ -11,6 +11,10 @@
 # To make modifications to this file, edit `tools/run_python_compiler.sh` and
 # then run `tools/create_entry_points.py`
 
+unset _PYTHON_SYSCONFIGDATA_NAME
+unset PYTHONHOME
+unset PYTHONPATH
```
Comment (sbc100) [inline em++:16]: Do you need these two lines if you are using `-E` below?

Comment (pmp-p) [reply em++:16]: i think -E does not clear the env it just disregard the values, so better safe than sorry later be these 2 are really critical if another python child process is launched without -E

Comment (sbc100) [reply em++:16]: OK, then in that case can you remove the `-E`?  There is no point in adding it if we have another method of doing that same thing and we think that we cannot inject `-E` into all the right place.  That would be `-E` as partial, and redundant solution here, right?

Maybe add a comment to that effect.  Something like "Unset these environment variables explicitly, rather than using `python -E`"

File: emranlib [hunk 3]
```diff
@@ -11,6 +11,10 @@
 # To make modifications to this file, edit `tools/run_python.sh` and then run
 # `tools/create_entry_points.py`
 
+unset _PYTHON_SYSCONFIGDATA_NAME
+unset PYTHONHOME
+unset PYTHONPATH
```
Comment (sbc100) [inline emranlib:16]: What about the windows .bat file launchers.. shouldn't we do that same there?

(I'm a little worried about windows since I'm not sure how scoping works in .bat files.. if you unset a variables does that unset it in the calling process too?)

File: em++ (hunk 2 above)
Comment (pmp-p) [reply em++:16]: -E covers (a lot!) of other variables that should not be impacted for emsdk, the two unset are really usefull failsafe for subprocesses where we I can't be sure -E will be set ( i did not read all emsdk source code).

tldr :
1) -E ensure fully proper operation of emsdk base tool called in first place.

2) the unset HOME/PATH ensure minimal proper operation of anything that could spawn as subprocesses of that emcc tool over those the shell script has no direct control.

File: emranlib (hunk 3 above)
Comment (pmp-p) [reply emranlib:16]: tbh i personnally don't intend to produce sysconfig data for windows->wasm cross compilation of CPython modules. i'd say just add only -E until somebody has a real problem with python subprocesses.

python-wasm-sdk primary target is ubuntu github CI, someone else would have to test that on windows if they really want to.

File: em++ (hunk 2 above)
Comment (sbc100) [reply em++:16]: Do you know what else `-E` covers?   I guess the doc is quite open ended:

```
       -E     Ignore environment variables like PYTHONPATH and PYTHONHOME that modify the  behavior
              of the interpreter.
```

There are indeed a couple of places where emscripten spawns internal stuff using `sys.interpreter`.  With this change, those would run in partially-clean environment rather then that fully cleaned `-E` one?    It seems odd that `-E` would not also cover python sub-processes.   It also seems odd that we would consider partially-cleaned OK in some circumstances but not OK in others.

One example of emcc launching a python subprocess: https://github.com/emscripten-core/emscripten/blob/4549d9da3e85c96a71a7cf56de5e203b22801bfb/tools/building.py#L1435-L1440

Comment (pmp-p) [reply em++:16]: >  It seems odd that -E would not also cover python sub-processes

i'm quite sure -E does not impact subprocesses and there's a lot of ways to spawn them from python or other children. if you can be 100% certain to not miss one them, then ensure running sys.executable with -E would be of course better option at emsdk level.

But there's still a high probability that a non python child program starts another python regardless of -E, relying only on unsetting PYTHON* : i'm thinking of emconfigure, emmake and emcmake

Comment (sbc100) [reply em++:16]: Does your use case really require `-E` or is removing those 3 environment variables sufficient to make it work?

Comment (pmp-p) [reply em++:16]: my case require both because we want to achieve good reproducibility if a user decide to build his own ci by downloading the sdk instead or using another CI than github, maybe want to upgrade emsdk without redownloading/rebuilding everything ( we don't want to patch like we do actualliy) , or use his custom python + custom env to drive emsdk that use his designed EMSDK_PYTHON.

Comment (sbc100) [reply em++:16]: So what happens if run a tool like `emconfigure`, `emmake`, `emcmake`?    Lets say I have a tool that is run at configure time, and its written in python.

If I run `PYTHONPATH=/required/by/mytool ./configure` I would expect the python tool within my configure script to honor the `PYTHONPATH` I think?   So if `emconfigure` sanitizing those environment variables won't that break my expectations?

It seems that non-emscripten python scripts probably do want to honor those paths, even though emscirpten itself wants to be hermetic?

Comment (sbc100) [reply em++:16]: Given that, I think the pure `-E` approach is probably best and we should work to ensure that emscripten tools are always launched via the launcher scripts (i've been thinking of adding a check for this actually to warn folks from inadvertently running `python emcc.py` themselves).

Comment (pmp-p) [reply em++:16]: Usually all system tools on good distribution should already set -E but again ... can't be 100% sure  ( it was worst in the past there are probably still remnants or obsolete systems that cannot upgrade for some reason )

PYTHONPATH or PYTHONHOME should never influence anything under configure/make/cmake ever and they should use a stable system python.

When the (auto)tools need to configure for **another python** they (should) provide a parameter to designate the targeted interpreter and run it for getting config data. This is where a given shell script wrapper should clear env then set PYTHON* and run the target correctly.

Comment (sbc100) [reply em++:16]: But `PYTHONPATH=/required/by/mytool ./configure` does work today for cases where configure run some python script as part of configuring. 

This change would break that no?  Same for `PYTHONPATH=/required/by/mytool emmake`.

It doesnt seem like it should be our job to sanitize the environment for all our subprocesses, especially in the case of emmake/emcmake/emconfigure where the subprocess are use controlled and the user has a reasonable expectation that the environment they set will be passed to the children of make/cmake/configure.

So... I suggest we start with just using `-E`.. since that is clearly correct, and revisit that if it turns out that doesn't solve your issue?

Comment (sbc100) [reply em++:16]: And maybe we should add `-E` to the windows `.bat` files for consistency?

Comment (pmp-p) [reply em++:16]: > But PYTHONPATH=/required/by/mytool ./configure does work today for cases where configure run some python script as part of configuring.

imho using PYTHON* that way is bad because configure is not a python script.

PYTHON* is for the end of the chain : the tool py script itself. not for influencing autotools that is thousands of lines that run thousands of tools .

> revisit that if it turns out that doesn't solve your issue

well actually it's not specifically my issue, it's more the various hard to solve potential issues that will come of thousands of modules around.

That we may hope to cross compile to wasm without patching emsdk ( pyodide still do that and iirc even rebuild it fully) and without having thousands of PR to sanitize subprocesses or helper tools we have not yet heard of.

 That in the case python-wasm-sdk come to interest more than just pygame ( less than  3minutes to build the wasm python module without docker, i'm sure it will get traction ! )


>  And maybe we should add -E to the windows .bat files for consistency?

i don't know anymore how system python runs on windows or mac systems (since 2.7), i think you should but i'm certainly not the one to ask if you can safely.

Comment (sbc100) [reply em++:16]: > imho using PYTHON* that way is bad because configure is not a python script.

It might be in-advised but that doesn't mean we should break anyone who want to use it.

> well actually it's not specifically my issue, it's more the various hard to solve potential issues that will come of thousands of modules around.
> 
> That we may hope to cross compile to wasm without patching emsdk ( pyodide still do that and iirc even rebuild it fully) and without having thousands of PR to sanitize subprocesses or helper tools we have not yet heard of.

I think that number of python subprocess that get run by emcc should be very limited, we should be able to keep track of them and ensure they all get run with -E.    

> i don't know anymore how system python runs on windows or mac systems (since 2.7), i think you should but i'm certainly not the one to ask if you can safely.

I think we should add `-E` everywhere (including on windows) for consistency.   It could be confusing otherwise.  We have a fair amount of CI testing to ensure it works, so you don't need to worry about locally testing it.

**Comment by pmp-p** [comment]
@hoodmane 
> . I think you should explicitly set up the environment you need and then emcc should pass it along. I think this is the way that environment variables normally work.

no. setup.py should naturally call the python in the PATH ( like in venv ), while all emsdk tools should use EMSDK_PYTHON ( which most likely is to the stable system python )

it's especially annoying to have 
```
build module (eg pymunk) =>
    python(clang) => setup.py
        setup.py => ./configure some C lib (eg libchipmunk)
            ./configure => emsdk and tools that fail by catching _PYTHON_SYSCONFIGDATA_NAME
```                                               
or to have to use the replay system from pyodide that use too much heuristic and is not always 32 bits aware.

**Comment by sbc100** [comment]
Ok, it sounds reasonable to leave `_PYTHON_SYSCONFIGDATA_NAME` in there then.. and have `-E` take a care of the rest in a less subprocess-invasive way.

Perhaps add a comment such as: `For some reason -E doesn't ignore _PYTHON_SYSCONFIGDATA_NAME like it does all the other python environment variables`.   (I wonder if its worth filing a an upstream bug about that?)

**Comment by pmp-p** [comment]
> (I wonder if its worth filing a an upstream bug about that?)

i think @tiran could say more on that he loves cross compiling ;) 

is check_call() is really the only other place -E is required  ?

do unset PYTHONHOME and unset PYTHONPATH go away ?

**Comment by sbc100** [comment]
> > (I wonder if its worth filing a an upstream bug about that?)
> 
> i think @tiran could say more on that he loves cross compiling ;)
> 
> is check_call() is really the only other place -E is required ?

I think it should be anywhere we use `shared.PYTHON` to run sub-processes.   And I think we should probably avoid ever doing that and instead use the wrapper scripts instead.

> 
> do unset PYTHONHOME and unset PYTHONPATH go away ?

Yes please.

**Review by sbc100 — APPROVED** [review]

**Review by sbc100 — APPROVED** [review]

**Comment by sbc100** [comment]
Do you mind if I re-title this PR something like `Run python with -E to ignore incoming PYTHONPATH/etc environment variables` when I land it?

**Comment by pmp-p** [comment]
please do as you wish
